data/data-pipeline/data_pipeline/etl/sources/census_acs_median_income/etl.py

```diff
@@ -0,0 +1,259 @@
+        )
+        self.MSA_ID_FIELD_NAME: str = "MSA ID"
+        self.MSA_TYPE_FIELD_NAME: str = "MSA Type"
+
+        # Set constants for MSA median incomes
+        self.MSA_MEDIAN_INCOME_URL: str = f"https://api.census.gov/data/{self.ACS_YEAR}/acs/acs5?get=B19013_001E&for=metropolitan%20statistical%20area/micropolitan%20statistical%20area"
+        self.MSA_INCOME_FIELD_NAME: str = f"Median household income in the past 12 months (MSA; {self.ACS_YEAR} inflation-adjusted dollars)"
+
+        # Set constants for state median incomes
+        self.STATE_MEDIAN_INCOME_URL: str = f"https://api.census.gov/data/{self.ACS_YEAR}/acs/acs5?get=B19013_001E&for=state"
+        self.STATE_GEOID_FIELD_NAME: str = "GEOID2"
+        self.STATE_MEDIAN_INCOME_FIELD_NAME: str = f"Median household income (State; {self.ACS_YEAR} inflation-adjusted dollars)"
+
+        # Constants for output
+        self.AMI_REFERENCE_FIELD_NAME: str = "AMI Reference"
+        self.AMI_FIELD_NAME: str = "Area Median Income (State or metropolitan)"
+
+    def _transform_geocorr(self) -> pd.DataFrame:
+        # Transform the geocorr data
+        geocorr_df = self.raw_geocorr_df
+
+        # Strip the unnecessary period from the tract ID:
+        geocorr_df["tract"] = geocorr_df["tract"].str.replace(
+            ".", "", regex=False
+        )
+
+        # Create the full GEOID out of the component parts.
+        geocorr_df[self.GEOID_FIELD_NAME] = (
+            geocorr_df["county"] + geocorr_df["tract"] + geocorr_df["bg"]
+        )
+
+        # QA the combined field:
+        tract_values = geocorr_df[self.GEOID_FIELD_NAME].str.len().unique()
+        if any(tract_values != [12]):
+            print(tract_values)
+            raise ValueError("Some of the census BG data has the wrong length.")
+
+        # Rename some fields
+        geocorr_df.rename(
+            columns={
+                "placenm": self.PLACE_FIELD_NAME,
+                "cbsaname10": self.MSA_FIELD_NAME,
+                "cntyname": self.COUNTY_FIELD_NAME,
+                "stab": self.STATE_ABBREVIATION_FIELD_NAME,
+                "cbsa10": self.MSA_ID_FIELD_NAME,
+                "cbsatype10": self.MSA_TYPE_FIELD_NAME,
+            },
+            inplace=True,
+            errors="raise",
+        )
+
+        # Remove duplicated rows.
+        # Some rows appear twice: once for the population within a CBG that's also within a census place,
+        # and once for the population that's within a CBG that's *not* within a census place.
+        # Drop the row that's not within a census place.
+
+        # Sort by whether the place has a place name:
+        geocorr_df.sort_values(
+            by=self.PLACE_FIELD_NAME, axis=0, ascending=True, inplace=True
+        )
+
+        # Drop all the duplicated rows except for the first one (which will have the place name):
+        rows_to_drop = geocorr_df.duplicated(
+            keep="first", subset=[self.GEOID_FIELD_NAME]
```

Comment:
I'm wary about logic that relies on order, even though we're the ones setting the order. I also see that we're ordering twice - first for the purpose of dropping dupes based on the order, and second to put the df back in its original order. Is it not possible to drop the dupes based on presence/absence of the place field name? It seems like that would make this faster and cleaner.

Reply:
Sadly `drop_duplicates` and `duplicated` in `pandas` only work using first or last, so they require a sort for this. 

We could look into a more elaborate methodology (like writing our own duplicates-finder code). Wdyt?

Reply:
Grr, pandas! If you think this is a conventional approach and not noticeably impacting performance, then I'm okay with it.